src/v/cloud_storage/remote_partition.cc

```diff
@@ -69,6 +71,104 @@ get_kafka_max_offset(const partition_manifest::segment_meta& m) {
+      key_offset);
+    _probe.segment_materialized();
+    return iter;
+}
+
+remote_partition::borrow_result_t remote_partition::borrow_next_reader(
+  storage::log_reader_config config, model::offset hint) {
+    // The code find the materialized that can satisfy the reader. If the
+    // segment is not materialized it materializes it.
+    // The following situations are possible:
+    // - The config.start_offset belongs to the same segment if
+    //   - Last record batches are config batches so config.start_offset
+    //     is <= segment.committed_offset.
+    //   - The underlying segment was reuploaded and now covers the wider
+    //     offset range. In this case we just stopped reading previous
+    //     version of the segment which now became longer.
+    // - The config.start_offset matches the base_offset - delta_offset of
+    //   the next segment.
+    auto kafka_offset_lookup = [this, hint](
+                                 const storage::log_reader_config& config) {
+        auto ko = model::offset_cast(config.start_offset);
+        // Two level lookup:
+        // - find segment meta based on kafka offset
+        //   this allow us to avoid any abiguity in case if the segment
+        //   doesn't have any data. The 'segment_containing' method of the
+        //   manifest takes this into account.
+        // - find materialized segment or materialize the new one
+        auto mit = _manifest.end();
+        if (hint == model::offset{}) {
+            // This code path is only used for the first lookup. It
+            // could be either lookup by kafka offset or by timestamp.
+            if (config.first_timestamp) {
+                auto maybe_meta = _manifest.timequery(*config.first_timestamp);
+                if (maybe_meta) {
+                    mit = _manifest.segment_containing(
+                      maybe_meta->get().base_offset);
+                }
+            } else {
+                // In this case the lookup is perfomed by kafka offset.
+                // Normally, this would be the first lookup done by the
+                // partition_record_batch_reader_impl. This lookup will
+                // skip segments without data batches (the logic is implemented
+                // inside the partition_manifest).
+                mit = _manifest.segment_containing(ko);
+            }
+            if (mit == _manifest.end()) {
+                // Segment that matches exactly can't be found
+                // in the manifest. In this case we want to start
+                // scanning from the begining of the partition if
+                // some conditions are met.
+                auto so = _manifest.get_start_offset().value_or(
+                  model::offset::min());
+                if (config.start_offset < so && config.max_offset > so) {
+                    mit = _manifest.begin();
+                }
+            }
+        } else {
+            mit = _manifest.segment_containing(hint);
```

Comment:
is it okay that we don't skip segments without data batches here?

Reply:
added that